web/client-api/src/main/java/io/deephaven/web/client/api/JsPartitionedTable.java

```diff
@@ -0,0 +1,202 @@
+public class JsPartitionedTable extends HasEventHandling {
+    public static final String EVENT_KEYADDED = "keyadded",
+            EVENT_DISCONNECT = JsTable.EVENT_DISCONNECT,
+            EVENT_RECONNECT = JsTable.EVENT_RECONNECT,
+            EVENT_RECONNECTFAILED = JsTable.EVENT_RECONNECTFAILED;
+
+    private final WorkerConnection connection;
+    private final JsWidget widget;
+    private List<String> keyColumnTypes;
+    private PartitionedTableDescriptor descriptor;
+    private JsTable keys;
+    private TableSubscription subscription;
+
+    /*
+     * Represents the sorta-kinda memoized results, tables that we've already locally fetched from the tablemap, and if
+     * all references to a table are released, entries here will be replaced with unresolved instances so we don't leak
+     * server references or memory. Keys are Object[], even with one element, so that we can easily hash without an
+     * extra wrapper object. Since columns are consistent with a PartitionedTable, we will not worry about "foo" vs
+     * ["foo"] as being different entries.
+     */
+    private final Map<List<Object>, JsLazy<Promise<ClientTableState>>> tables = new HashMap<>();
+
+
+    @JsIgnore
+    public JsPartitionedTable(WorkerConnection connection, JsWidget widget) {
+        this.connection = connection;
+        this.widget = widget;
+    }
+
+    @JsIgnore
+    public Promise<JsPartitionedTable> refetch() {
+        if (keys != null) {
+            keys.close();
+        }
+        if (subscription != null) {
+            subscription.close();
+        }
+        return widget.refetch().then(w -> {
+            descriptor = PartitionedTableDescriptor.deserializeBinary(w.getDataAsU8());
+
+            keyColumnTypes = new ArrayList<>();
+            ColumnDefinition[] columnDefinitions = BarrageUtils.readColumnDefinitions(
+                    BarrageUtils.readSchemaMessage(descriptor.getConstituentDefinitionSchema_asU8()));
+            for (int i = 0; i < columnDefinitions.length; i++) {
+                ColumnDefinition columnDefinition = columnDefinitions[i];
+                if (descriptor.getKeyColumnNamesList().indexOf(columnDefinition.getName()) != -1) {
+                    keyColumnTypes.add(columnDefinition.getType());
+                }
+            }
+
+            return w.getExportedObjects()[0].fetch();
+        }).then(result -> {
+            keys = (JsTable) result;
+            subscription = keys.subscribe(
+                    JsArray.asJsArray(keys.findColumns(descriptor.getKeyColumnNamesList().asArray(new String[0]))));
+            subscription.addEventListener(TableSubscription.EVENT_UPDATED, this::handleKeys);
+
+            LazyPromise<JsPartitionedTable> promise = new LazyPromise<>();
+            subscription.addEventListenerOneShot(TableSubscription.EVENT_UPDATED, data -> promise.succeed(this));
+            keys.addEventListener(JsTable.EVENT_DISCONNECT, promise::fail);
```

Comment:
Maybe fail with an appropriate error message here... and do you need to close the `subscription` here?

Reply:
Disconnect implies that we lost contact with the server in a way we couldnt fix, so the subscription is dead too.

Note that disconnect/reconnect is not implemented at all in DHC.